pkg/agent/openflow/multicast.go

```diff
@@ -99,6 +132,82 @@ func (f *featureMulticast) multicastOutputFlow(cookieID uint64) binding.Flow {
+		// group and its members in the meanwhile.
+		// Do not set dst IP address because IGMPv1 report message uses target multicast group as IP destination in
+		// the packet.
+		table.ofTable.BuildFlow(priorityHigh).
+			Cookie(f.cookieAllocator.Request(f.category).Raw()).
+			MatchProtocol(binding.ProtocolIGMP).
+			MatchRegMark(FromLocalRegMark).
+			Action().LoadRegMark(CustomReasonIGMPRegMark).
+			Action().SendToController(reason).
+			Action().Normal().
+			Done(),
+	}
+	return flows
+}
+
+// localMulticastForwardFlows generates the flow to forward multicast packets with OVS action "normal", and output
+// it to Antrea gateway in the meanwhile, so that the packet can be forwarded to local Pods which have joined the multicast
+// group and to the external receivers. For multicast packets from the external network, they also hit the flow and the packets
+// are not sent back to Antrea gateway, because OVS datapath will drop them when it finds the output port is the same as the input port.
+func (f *featureMulticast) localMulticastForwardFlows(multicastIP net.IP, groupID binding.GroupIDType) []binding.Flow {
+	return []binding.Flow{
+		MulticastRoutingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(f.cookieAllocator.Request(f.category).Raw()).
+			MatchProtocol(binding.ProtocolIP).
+			MatchDstIP(multicastIP).
+			Action().Group(groupID).
+			Done(),
+	}
+}
+
+// multicastEgressPodMetricFlows generates the flows to forward multicast egress packets before
+// outputting to MulticastOutputTable. It matches source IP with IP of the local sender Pod.
+func (f *featureMulticast) multicastEgressPodMetricFlows(podInterfaceIP net.IP) binding.Flow {
+	ipProtocol := getIPProtocol(podInterfaceIP)
+	return MulticastEgressPodMetricTable.ofTable.BuildFlow(priorityNormal).
+		MatchProtocol(ipProtocol).
+		MatchSrcIP(podInterfaceIP).
+		MatchDstIPNet(*types.McastCIDR).
+		Action().NextTable().
+		Done()
+}
+
+// externalMulticastReceiverFlow generates the flow to output multicast packets to Antrea gateway, so that local Pods can
+// send multicast packets to the external receivers. For the case that one or more local Pods have joined the target
+// multicast group, it is handled by the flows created by function "localMulticastForwardFlows" after local Pods report the
+// IGMP membership.
+func (f *featureMulticast) externalMulticastReceiverFlow() binding.Flow {
+	return MulticastRoutingTable.ofTable.BuildFlow(priorityLow).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchProtocol(binding.ProtocolIP).
+		MatchDstIPNet(*types.McastCIDR).
+		Action().LoadRegMark(OFPortFoundRegMark).
+		Action().LoadToRegField(TargetOFPortField, config.HostGatewayOFPort).
+		Action().GotoTable(MulticastOutputTable.GetID()).
+		Done()
+}
+
+func (f *featureMulticast) multicastIngressIGMPPodMetricFlow() binding.Flow {
+	return MulticastIngressPodMetricTable.ofTable.BuildFlow(priorityHigh).
```

Comment:
If this is for IGMP query packet, it should match IGMP protocol. And if you want to collect ingress metrics per Pod, it should match Pod ofport number with reg1

Reply:
updated. This is for IGMP query packet

Reply:
According to the offline discussion, this flow is actually used to bypass IGMP query packet in MulticastIngressPodMetricTable, but not to collect the IGMP query per Pod metrics. By now, the IGMP Query metrics is performed per ANP/ACNP rule, but not collected per Pod. Would you add comments to explain this?

Reply:
Comment added in InstallMulticastInitialFlows